### DISCUSSION TITLE
RuntimeError: [E039] Array bounds exceeded while searching for root word while using own .conllu to spacy doc converter

### DISCUSSION THREAD

**Discussion by ghost** [explosion/spaCy#8296] [Help: Coding & Implementations] [answered]
Hey. I encountered a problem with visualization via displacy while using spacy doc object generated from conllu file. Here is the full traceback:

```
vis_conllu(r'C:\Other Developments\conllu_1-100\new_2.conllu')
vis_conlluTraceback (most recent call last):
  Python Shell, prompt 2, line 1
    # Used internally for debug sandbox under external interpreter
  File "C:\Users\G. Kostenko\AppData\Local\Programs\Python\Python37\Lib\site-packages\ganalyzer.py", line 1363, in vis_conllu
    displacy.serve(doc, style="dep")
  File "C:\Users\G. Kostenko\AppData\Local\Programs\Python\Python37\Lib\site-packages\spacy\displacy\__init__.py", line 96, in serve
    render(docs, style=style, page=page, minify=minify, options=options, manual=manual)
  File "C:\Users\G. Kostenko\AppData\Local\Programs\Python\Python37\Lib\site-packages\spacy\displacy\__init__.py", line 53, in render
    parsed = [converter(doc, options) for doc in docs] if not manual else docs
  File "C:\Users\G. Kostenko\AppData\Local\Programs\Python\Python37\Lib\site-packages\spacy\displacy\__init__.py", line 53, in <listcomp>
    parsed = [converter(doc, options) for doc in docs] if not manual else docs
  File "C:\Users\G. Kostenko\AppData\Local\Programs\Python\Python37\Lib\site-packages\spacy\displacy\__init__.py", line 148, in parse_deps
    retokenizer.merge(span, attrs=attrs)
  File "C:\Users\G. Kostenko\AppData\Local\Programs\Python\Python37\Lib\site-packages\spacy\tokens\_retokenize.cp37-win_amd64.pyd", line 120, in spacy.tokens._retokenize.Retokenizer.__exit__
    H)uHAÿP0H
ÿtH/u
  File "C:\Users\G. Kostenko\AppData\Local\Programs\Python\Python37\Lib\site-packages\spacy\tokens\_retokenize.cp37-win_amd64.pyd", line 181, in spacy.tokens._retokenize._merge
    HÕAÿÖ
Àu_H\000\000\000H
Ét
  File "C:\Users\G. Kostenko\AppData\Local\Programs\Python\Python37\Lib\site-packages\spacy\tokens\span.cp37-win_amd64.pyd", line 578, in spacy.tokens.span.Span.root.__get__
    HCHËÿP0H
  File "C:\Users\G. Kostenko\AppData\Local\Programs\Python\Python37\Lib\site-packages\spacy\tokens\span.cp37-win_amd64.pyd", line 760, in spacy.tokens.span._count_words_to_root
    Ù¦\000è|ÊÿÿHm§\000H
ÀÙ\000\000\000H
builtins.RuntimeError: [E039] Array bounds exceeded while searching for root word. This likely means the parse tree is in an invalid state. Please report this issue here: http://github.com/explosion/spaCy/issues
```

Obviously, the problem is caused by the way I'm creating a doc object. I asked how to do that some time ago, and then slightly modified the proposed code.

- GitHub Discussions: https://github.com/explosion/spaCy/issues/5600

So, here is how I try to visualize:

```
def vis_conllu(conllu_dir):
    
    print('vis_conllu')
    
    with open(conllu_dir, encoding="UTF-8") as file:
        string = file.read()
    
    doc = from_conllu(string)
    
    displacy.serve(doc, style="dep")
```

Here is what `from_conllu` means:

```
def from_conllu(conllu_string):
    
    nlp = spacy.blank("de")
    
    doc = doc_from_conllu(nlp.vocab, conllu_string)
    
    return doc
```

Here is how I get doc from conllu:

```
import spacy
from spacy.tokens import Doc

def doc_from_conllu(vocab, raw_conllu):
    
    words, spaces, tags, poses, morphs, lemmas = [], [], [], [], [], []
    heads, deps = [], []
    
    lines = raw_conllu.split('\n')
    lines = [i for i in lines if len(i) is not 0 and not i.startswith('#') and i not in ['\r']]
    
    token_id = 0
    
    for i in range(len(lines)):
        line = lines[i]
        parts = line.split("\t")
        original_id, word, lemma, pos, tag, morph, head, dep, _1, misc = parts
        if "." in original_id or "-" in original_id:
            continue
        if "SpaceAfter=No" in misc:
            spaces.append(False)
        else:
            spaces.append(True)

        head = token_id + (int(head) - int(original_id))
        tag = pos if tag == "_" else tag
        morph = morph if morph != "_" else ""
        dep = "ROOT" if dep == "root" else dep

        words.append(word)
        lemmas.append(lemma)
        poses.append(pos)
        tags.append(tag)
        morphs.append(morph)
        heads.append(head)
        deps.append(dep)
        
        token_id += 1

    doc = Doc(vocab, words=words, spaces=spaces)
    
    for i in range(len(doc)):
        doc[i].tag_ = tags[i]          
        doc[i].pos_ = poses[i]
        doc[i].dep_ = deps[i]
        doc[i].lemma_ = lemmas[i]
        doc[i].head = doc[heads[i]]
    doc.is_parsed = True
    doc.is_tagged = True

    return doc
```

... and where, I guess, the problem originates from. Could you advise why the root cannot be located? Is it somehow connected to the sent_id?.. And how to fix it?

- Operating System: WIn10
- Python Version Used: 3.7

**Comment by polm** [comment, marked as answer]
Can you find an example sentence with issues and print out the head values you're setting? That's probably the easiest way to see what's going wrong. 

    head = token_id + (int(head) - int(original_id))

Looking at this line, it looks like you'll end up with roughly the value of `int(head)` here. That might be a relative token index (like -1 means "the token before this one is the head"), which will cause errors based on how you're assigning things - you need the index in the doc. If you print your values this should be easy to check.

Also, to be clear, are you using spaCy v2? 

As another minor point, when using blocks of code in Markdown you need to use three backticks on a line before and after the code, not one at the start and end.

**Reply by ghost** [reply]
Thank you! Somehow, the error is related to punctuation. If I take this sentence:

![Скриншот 08-06-2021 14 21 36](https://user-images.githubusercontent.com/58014330/121176628-1b49b700-c865-11eb-90f8-632a4f5a85df.png)

The visualization returns an error like above.

However, if  I delete the last line:

![Скриншот 08-06-2021 14 24 41](https://user-images.githubusercontent.com/58014330/121176767-46340b00-c865-11eb-89a1-51c27768e167.png)


The visualization works just fine.

I'm using spacy 2.3.5

Update: if I just change "." in the last line to "point", it works as well. Somehow, "." is not recognized as word.

**Reply by polm** [reply]
I have been looking at your code for a while and I think I figured out what's wrong. Your head calculation appears to be wrong, this worked for me:

```
head = int(head)
if head == 0: head = token_id + 1 # because 0 marks the root
head = head - 1 # because we need 0 base token ids, but conllu is 1-base
```

Additionally, your punctuation related problems seem to come from a combination of your head calculation issue and displaCy merging punctuation by default. You can turn off punctuation merging like this:

    displacy.serve(doc, style="dep", options={"collapse_punct": False})

But if you fix your head calculation you shouldn't have to.

Also, as a note for the future, please post your data in text form, not screenshots.

**Reply by ghost** [reply]
Thanks a lot, it solved the problem.
